### PR TITLE
[dbnode] Add configuration to flush data in parallel for peer streaming

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

This allows for faster progression if bottlenecked by disk flush processing time when only flushing single shard at a time to disk.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/dbnode/client/session.go [hunk 1]
```diff
@@ -3629,7 +3627,7 @@ type enqueueCh struct {
 	metrics              *streamFromPeersMetrics
 }
 
-const enqueueChannelDefaultLen = 32768
+const enqueueChannelDefaultLen = 512
```
Comment (yyinsc) [inline src/dbnode/client/session.go:3630]: hmmm? isn't this too small in a sense?

File: src/dbnode/storage/index/convert/convert.go [hunk 2]
```diff
@@ -211,7 +212,13 @@ func TagsFromTagsIter(
 			idRef = true
 		} else {
 			if idPool != nil {
-				tag.Name = idPool.Clone(curr.Name)
+				// NB(r): Fast path for if a graphite tag name to save
+				// a lot of space is to reuse a preallocated tag name.
+				if idx, ok := graphite.TagIndex(nameBytes); ok {
```
Comment (yyinsc) [inline src/dbnode/storage/index/convert/convert.go:217]: Might worth either put this change in a separate PR or call it out. this is a very nice change with decent cpu reduction.

**Review by yyinsc — APPROVED** [review]

File: src/dbnode/client/session.go (hunk 1 above)
Comment (robskillington) [reply src/dbnode/client/session.go:3635]: Since the enqueue channel just moves things from one incoming channel into requests that go into a host queue, it's never blocked by the network or anything like that so it actually doesn't need a long queue at all (and each item does not require a lot of computation to complete).

I was seeing some non-trivial memory usage in the profiles from this channel since each entry contains data from many replicas and is a very large struct (several hundreds of bytes). So since this was never seeing any lag I think making this small shouldn't hurt anything as it is just shuffling things between other queues and is not blocked by any network/other dependencies.

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go [hunk 3]
```diff
@@ -619,38 +630,45 @@ func (s *peersSource) flush(
...
-	for _, entry := range shardResult.AllSeries().Iter() {
-		series := entry.Value()
-		numBlocksRemaining := len(series.Blocks.AllBlocks())
-		// Should never happen since we removed all the block in the previous loop and fetching
-		// bootstrap blocks should always be exclusive on the end side.
-		if numBlocksRemaining > 0 {
-			numSeriesTriedToRemoveWithRemainingBlocks++
-			continue
-		}
+	// Perform cleanup async but allow caller to wait on them.
```
Comment (ryanhall07) [inline src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:633]: nice. makes sense

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go [hunk 4]
```diff
@@ -567,43 +596,25 @@ func (s *peersSource) flush(
 				continue
 			}
 
-			flushCtx.Reset()
-			stream, err := bl.Stream(flushCtx)
+			checksum, err := bl.Checksum()
```
Comment (ryanhall07) [inline src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:599]: I don't follow the impact of these 3 changes that seem to reorder operations a bit

**Review by ryanhall07 — APPROVED** [review]
LG

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go [hunk 5]
```diff
@@ -273,13 +256,14 @@ func (s *peersSource) readData(
...
 		resultLock              sync.Mutex
 		wg                      sync.WaitGroup
-		persistenceWorkerDoneCh = make(chan struct{})
 		persistenceMaxQueueSize = s.opts.PersistenceMaxQueueSize()
 		persistenceQueue        = make(chan persistenceFlush, persistenceMaxQueueSize)
 		resultOpts              = s.opts.ResultOptions()
 		count                   = shardTimeRanges.Len()
 		concurrency             = s.opts.DefaultShardConcurrency()
 		blockSize               = nsMetadata.Options().RetentionOptions().BlockSize()
+		persistWg               = &sync.WaitGroup{}
```
Comment (nbroyles) [inline src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:265]: nit: sync.WaitGroup is more idiomatic here, yes?

Comment (nbroyles) [reply src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:265]: Oh, I get it. This is so you get the actual reference. Makes sense.

Comment (ryanhall07) [reply src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:265]: you mean without the ref? since it's being passed around between functions you need the ref

Comment (nbroyles) [reply src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:265]: I think we're intending to say the same thing, but basically he's creating a pointer so the sync.WaitGroup object itself doesn't get copied when passed to the function using it.

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go (hunk 4 above)
Comment (nbroyles) [reply src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:599]: +1

**Review by nbroyles — APPROVED** [review]

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go (hunk 5 above)
Comment (ryanhall07) [reply src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:265]: lol i think we commented at the same time

File: src/dbnode/client/session.go [hunk 6]
```diff
@@ -68,13 +68,11 @@ import (
...
-	clusterConnectWaitInterval           = 10 * time.Millisecond
-	blocksMetadataChannelInitialCapacity = 4096
-	gaugeReportInterval                  = 500 * time.Millisecond
-	blockMetadataChBufSize               = 4096
-	shardResultCapacity                  = 4096
-	hostNotAvailableMinSleepInterval     = 1 * time.Millisecond
-	hostNotAvailableMaxSleepInterval     = 100 * time.Millisecond
+	clusterConnectWaitInterval       = 10 * time.Millisecond
+	gaugeReportInterval              = 500 * time.Millisecond
+	blockMetadataChBufSize           = 65536
```
Comment (arnikola) [inline src/dbnode/client/session.go:73]: That's quite a substantial increase, is this intentional?

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source_index_test.go [hunk 7]
```diff
@@ -468,7 +468,7 @@ func TestBootstrapIndex(t *testing.T) {
 //			_ result.Options,
 //		) (result.ShardResult, error) {
 //			goodID := ident.StringID("foo")
-//			goodResult := result.NewShardResult(0, opts.ResultOptions())
+//			goodResult := result.NewShardResult(opts.ResultOptions())
```
Comment (arnikola) [inline src/dbnode/storage/bootstrap/bootstrapper/peers/source_index_test.go:471]: nit: should this commented section just be deleted?

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go [hunk 8]
```diff
@@ -290,8 +274,16 @@ func (s *peersSource) readData(
...
 		zap.Bool("shouldPersist", shouldPersist))
 	if shouldPersist {
-		go s.startPersistenceQueueWorkerLoop(
-			opts, persistenceWorkerDoneCh, persistenceQueue, persistFlush, result, &resultLock)
+		// Spin up persist workers.
+		for i := 0; i < s.opts.ShardPersistenceFlushConcurrency(); i++ {
+			closer, err := s.startPersistenceQueueWorkerLoop(opts,
+				persistWg, persistenceQueue, result, &resultLock)
+			if err != nil {
+				return nil, err
```
Comment (arnikola) [inline src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:282]: Should this stop the other workers?

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go (hunk 4 above)
Comment (arnikola) [reply src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:599]: +2

File: src/dbnode/integration/peers_bootstrap_high_concurrency_test.go [hunk 9]
```diff
@@ -23,33 +23,71 @@
...
+	"github.com/m3db/m3/src/x/ident"
 	xtest "github.com/m3db/m3/src/x/test"
 
 	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
 )
 
 func TestPeersBootstrapHighConcurrency(t *testing.T) {
+	for _, test := range []testPeersBootstrapHighConcurrencyOptions{
+		{BatchSize: 16, Concurrency: 64, BatchesPerWorker: 8},
```
Comment (arnikola) [inline src/dbnode/integration/peers_bootstrap_high_concurrency_test.go:45]: Nice, were there other configs worth trying?

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go (hunk 4 above)
Comment (robskillington) [reply src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:604]: So now we aren't doing `.Stream()` anymore and we are discarding the block (which is good because `.Stream()` is expensive), we need to do `.Checksum()` before (the block will already have the pre-calculated checksum).

File: src/dbnode/client/session.go (hunk 6 above)
Comment (robskillington) [reply src/dbnode/client/session.go:73]: Yes, this was unfortunately a significant bottleneck since IO work is required to pull off the queue here and block the producer frequently. Also the individual channel element struct is rather small so this doesn't add much overall space / usage.

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source_index_test.go (hunk 7 above)
Comment (robskillington) [reply src/dbnode/storage/bootstrap/bootstrapper/peers/source_index_test.go:472]: Sure thing.

File: src/dbnode/storage/bootstrap/bootstrapper/peers/source.go (hunk 8 above)
Comment (robskillington) [reply src/dbnode/storage/bootstrap/bootstrapper/peers/source.go:287]: The inner workers all take work from the `persistenceQueue` channel which gets closed later `close(persistenceQueue)` (and stops all the workers) and then it waits on these to finish their cleanups with `persistWg.Wait()`.

File: src/dbnode/integration/peers_bootstrap_high_concurrency_test.go (hunk 9 above)
Comment (robskillington) [reply src/dbnode/integration/peers_bootstrap_high_concurrency_test.go:45]: The tests take a substantial time, so perhaps I'll add just one more with a higher batch size and lower concurrency, maybe invert these numbers.

Comment (robskillington) [reply src/dbnode/integration/peers_bootstrap_high_concurrency_test.go:45]: Added one inverted test.

**Review by arnikola — APPROVED** [review]
LGTM
